toolchain/check/import_ref.cpp

```diff
@@ -196,6 +204,66 @@ class ImportRefResolver {
     bool finished = true;
   };
 
+  auto FindResolvedConstId(SemIR::InstId inst_id)
+      -> std::pair<SemIR::ConstantId, llvm::SmallVector<SemIR::ImportIRInst>> {
+    if (auto existing_const_id = import_ir_constant_values().Get(inst_id);
+        existing_const_id.is_valid()) {
+      return {existing_const_id, {}};
+    }
+
+    const auto* cursor_ir = &import_ir_;
+    auto cursor_ir_id = SemIR::ImportIRId::Invalid;
+    auto cursor_inst_id = inst_id;
+
+    llvm::SmallVector<SemIR::ImportIRInst> indirect_insts;
+
+    while (true) {
+      auto loc_id = cursor_ir->insts().GetLocId(cursor_inst_id);
+      if (!loc_id.is_import_ir_inst_id()) {
+        return {SemIR::ConstantId::Invalid, indirect_insts};
+      }
+      auto ir_inst =
+          cursor_ir->import_ir_insts().Get(loc_id.import_ir_inst_id());
+
+      const auto* prev_ir = cursor_ir;
+      auto prev_inst_id = cursor_inst_id;
+
+      cursor_ir = cursor_ir->import_irs().Get(ir_inst.ir_id).sem_ir;
+      cursor_ir_id = context_.check_ir_map()[cursor_ir->check_ir_id().index];
+      if (!cursor_ir_id.is_valid()) {
+        // TODO: Should we figure out a location to assign here?
+        cursor_ir_id = AddImportIR(
+            context_, {.node_id = Parse::NodeId::Invalid, .sem_ir = cursor_ir});
+      }
+      cursor_inst_id = ir_inst.inst_id;
+
+      CARBON_CHECK(cursor_ir != prev_ir || cursor_inst_id != prev_inst_id)
+          << cursor_ir->insts().Get(cursor_inst_id);
+
+      if (auto const_id =
+              context_.import_ir_constant_values()[cursor_ir_id.index].Get(
+                  cursor_inst_id);
+          const_id.is_valid()) {
+        SetResolvedConstId(inst_id, indirect_insts, const_id);
+        return {const_id, {}};
+      } else {
+        indirect_insts.push_back(
+            {.ir_id = cursor_ir_id, .inst_id = cursor_inst_id});
+      }
+    }
+  }
```

Comment:
The `return`s in this function are going to perform a copy of the small vector. Using `std::move` doesn't entirely help because `SmallVector`'s inline storage still needs to be copied. Consider using an NRVO style instead: declare a function-scope variable whose type is the return type of the function and make all the `return` statements return that variable directly.
```suggestion
  auto FindResolvedConstId(SemIR::InstId inst_id)
      -> std::pair<SemIR::ConstantId, llvm::SmallVector<SemIR::ImportIRInst>> {
    std::pair<SemIR::ConstantId, llvm::SmallVector<SemIR::ImportIRInst>> result = {SemIR::ConstantId::Invalid, {}};
    if (auto existing_const_id = import_ir_constant_values().Get(inst_id);
        existing_const_id.is_valid()) {
      result.first = existing_const_id;
      return result;
    }

    const auto* cursor_ir = &import_ir_;
    auto cursor_ir_id = SemIR::ImportIRId::Invalid;
    auto cursor_inst_id = inst_id;

    while (true) {
      auto loc_id = cursor_ir->insts().GetLocId(cursor_inst_id);
      if (!loc_id.is_import_ir_inst_id()) {
        return result;
      }
      auto ir_inst =
          cursor_ir->import_ir_insts().Get(loc_id.import_ir_inst_id());

      const auto* prev_ir = cursor_ir;
      auto prev_inst_id = cursor_inst_id;

      cursor_ir = cursor_ir->import_irs().Get(ir_inst.ir_id).sem_ir;
      cursor_ir_id = context_.check_ir_map()[cursor_ir->check_ir_id().index];
      if (!cursor_ir_id.is_valid()) {
        // TODO: Should we figure out a location to assign here?
        cursor_ir_id = AddImportIR(
            context_, {.node_id = Parse::NodeId::Invalid, .sem_ir = cursor_ir});
      }
      cursor_inst_id = ir_inst.inst_id;

      CARBON_CHECK(cursor_ir != prev_ir || cursor_inst_id != prev_inst_id)
          << cursor_ir->insts().Get(cursor_inst_id);

      if (auto const_id =
              context_.import_ir_constant_values()[cursor_ir_id.index].Get(
                  cursor_inst_id);
          const_id.is_valid()) {
        SetResolvedConstId(inst_id, result.second, const_id);
        result = {const_id, {}};
        return result;
      } else {
        result.second.push_back(
            {.ir_id = cursor_ir_id, .inst_id = cursor_inst_id});
      }
    }
  }
```

Reply:
Done, making a struct to avoid repetition. Also made me notice to add comments.